File: data-structures-array-processing-advanced/src/main/java/com/arraymanipulation/utils/ArrayUtils.java
```java
package com.arraymanipulation.utils;

import java.util.Arrays;
import java.util.Random;

/**
 * Utility class for array-related operations, such as printing,
 * copying, and generating random arrays.
 */
public class ArrayUtils {

    private static final Random random = new Random();

    /**
     * Prints an integer array to the console in a readable format.
     *
     * @param array The integer array to print.
     * @param prefix An optional prefix string to print before the array.
     */
    public static void printArray(int[] array, String prefix) {
        System.out.println(prefix + ": " + (array == null ? "null" : Arrays.toString(array)));
    }

    /**
     * Prints an integer array to the console with a default "Array" prefix.
     *
     * @param array The integer array to print.
     */
    public static void printArray(int[] array) {
        printArray(array, "Array");
    }

    /**
     * Creates a deep copy of an integer array.
     * This is useful for testing or when multiple operations need to be performed
     * on the original state of an array.
     *
     * @param original The array to copy.
     * @return A new array containing the same elements as the original,
     *         or null if the original array is null.
     */
    public static int[] deepCopy(int[] original) {
        if (original == null) {
            return null;
        }
        return Arrays.copyOf(original, original.length);
    }

    /**
     * Generates a random integer array.
     *
     * @param size The desired size of the array.
     * @param minVal The minimum possible value for elements (inclusive).
     * @param maxVal The maximum possible value for elements (exclusive).
     * @return A new integer array with random values.
     * @throws IllegalArgumentException if size is negative or maxVal is not greater than minVal.
     */
    public static int[] generateRandomArray(int size, int minVal, int maxVal) {
        if (size < 0) {
            throw new IllegalArgumentException("Array size cannot be negative.");
        }
        if (maxVal <= minVal) {
            throw new IllegalArgumentException("maxVal must be greater than minVal.");
        }

        int[] array = new int[size];
        for (int i = 0; i < size; i++) {
            array[i] = random.nextInt(maxVal - minVal) + minVal;
        }
        return array;
    }

    /**
     * Checks if two integer arrays are equal in content and order.
     *
     * @param arr1 The first array.
     * @param arr2 The second array.
     * @return true if arrays are equal, false otherwise.
     */
    public static boolean areArraysEqual(int[] arr1, int[] arr2) {
        return Arrays.equals(arr1, arr2);
    }
}

```

File: optimization-array-processing-problems-enhanced/src/utils/arrayUtils.js
```javascript
/**
 * src/utils/arrayUtils.js
 *
 * A collection of utility functions for array manipulation and comparison,
 * useful for testing and benchmarking.
 */

/**
 * Compares two arrays for deep equality.
 * Handles nested arrays and primitive values.
 *
 * @param {Array<any>} arr1 The first array.
 * @param {Array<any>} arr2 The second array.
 * @returns {boolean} True if arrays are deeply equal, false otherwise.
 */
function areArraysDeepEqual(arr1, arr2) {
    if (arr1 === arr2) return true;
    if (arr1 == null || arr2 == null) return false;
    if (arr1.length !== arr2.length) return false;

    for (let i = 0; i < arr1.length; i++) {
        const val1 = arr1[i];
        const val2 = arr2[i];

        if (Array.isArray(val1) && Array.isArray(val2)) {
            if (!areArraysDeepEqual(val1, val2)) {
                return false;
            }
        } else if (val1 !== val2) {
            return false;
        }
    }
    return true;
}

/**
 * Creates a shallow copy of an array.
 * Useful for functions that modify arrays in place, to preserve original data for tests.
 *
 * @param {Array<any>} arr The array to copy.
 * @returns {Array<any>} A new array with the same elements as `arr`.
 */
function shallowCopyArray(arr) {
    return [...arr];
}

/**
 * Generates a random array of integers within a specified range.
 *
 * @param {number} size The desired size of the array.
 * @param {number} min The minimum value for elements (inclusive).
 * @param {number} max The maximum value for elements (inclusive).
 * @returns {number[]} An array of random integers.
 */
function generateRandomArray(size, min, max) {
    const arr = [];
    for (let i = 0; i < size; i++) {
        arr.push(Math.floor(Math.random() * (max - min + 1)) + min);
    }
    return arr;
}

/**
 * Generates a random array suitable for `findDuplicate` problems.
 * Contains `n+1` numbers in range `[1, n]` with exactly one duplicate.
 *
 * @param {number} n The upper bound for numbers and `n` in `n+1` length.
 * @returns {number[]} An array with a single duplicate.
 */
function generateArrayWithOneDuplicate(n) {
    const arr = [];
    for (let i = 1; i <= n; i++) {
        arr.push(i);
    }
    // Add a duplicate of a random number from 1 to n
    const duplicateValue = Math.floor(Math.random() * n) + 1;
    arr.push(duplicateValue);

    // Shuffle the array to randomize position of duplicate
    for (let i = arr.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [arr[i], arr[j]] = [arr[j], arr[i]];
    }
    return arr;
}

/**
 * Checks if an array is sorted in non-decreasing order.
 *
 * @param {Array<number>} arr The array to check.
 * @returns {boolean} True if sorted, false otherwise.
 */
function isSorted(arr) {
    for (let i = 0; i < arr.length - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

module.exports = {
    areArraysDeepEqual,
    shallowCopyArray,
    generateRandomArray,
    generateArrayWithOneDuplicate,
    isSorted,
};
```